src/System.Management.Automation/engine/SessionStateLocationAPIs.cs

```diff
@@ -205,23 +205,26 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context)
             ProviderInfo provider = null;
             string providerId = null;
 
-            // Replace path with last working directory when '-' was passed.
-            bool pushNextLocation = true;
-            if (originalPath.Equals("-", StringComparison.OrdinalIgnoreCase))
+            switch (originalPath)
             {
-                if (_SetLocationHistory.Count <= 0)
-                {
-                    throw new InvalidOperationException(SessionStateStrings.SetContentToLastLocationWhenHistoryIsEmpty);
-                }
-                var previousLocation =  _SetLocationHistory.Pop();
-                path = previousLocation.Path;
-                pushNextLocation = false;
-            }
-
-            if (pushNextLocation)
-            {
-                var newPushPathInfo = GetNewPushPathInfo();
-                _SetLocationHistory.Push(newPushPathInfo);
+                case var originalPathSwitch when originalPathSwitch.Equals("-", StringComparison.OrdinalIgnoreCase):
+                    if (_SetLocationHistory.UndoCount <= 0)
+                    {
+                        throw new InvalidOperationException(SessionStateStrings.LocationUndoStackIsEmpty);
+                    }
+                    path = _SetLocationHistory.Undo(this.CurrentLocation).Path;
+                    break;
+                case var originalPathSwitch when originalPathSwitch.Equals("+", StringComparison.OrdinalIgnoreCase):
+                    if (_SetLocationHistory.RedoCount <= 0)
+                    {
+                        throw new InvalidOperationException(SessionStateStrings.LocationRedoStackIsEmpty);
+                    }
+                    path = _SetLocationHistory.Redo(this.CurrentLocation).Path;
+                    break;
+                default:
+                    var newPushPathInfo = GetNewPushPathInfo();
```

Comment:
Ok, I renamed it to `pushPathInfo` instead, which is the name of the type here. which seemed to be the better compromise after looking at it again (otherwise the word `pushpathInfo` would've been repeated 3 times in one line)